File: PCorpKit/Sources/PCorpKit/ScheduledNotifications.swift
import Foundation
import UserNotifications

/// Real future-dated device notifications (e.g. "wake me up at 6am") --
/// distinct from SystemNotification.swift's immediate, in-session pushes
/// (used for things like a save_memory confirmation). Talks to
/// UNUserNotificationCenter directly on both platforms rather than
/// through a pluggable per-platform poster: unlike immediate pushes,
/// which historically needed osascript on desktop because a raw `swift
/// build` executable has no bundle identity and UNUserNotificationCenter
/// hard-crashed on it ("bundleProxyForCurrentProcess is nil"),
/// UserNotifications is genuinely cross-platform and confirmed live
/// (2026-09-18) to work fine now that build_app.sh assembles a real,
/// signed .app bundle.
///
/// The property that actually makes this useful: once scheduled, this is
/// registered with the OS's own notification scheduler, so it fires even
/// if the app isn't open or connected to the backend at the target time --
/// unlike the websocket-relayed `[notify]` sentinel, which only fires
/// while a live connection happens to be open. Real Apple platform
/// restriction, not a P Corp scoping choice: no third-party app can
/// create an actual system Alarm (Clock app) on either iOS or macOS, so
/// this is the closest working substitute -- a real banner + sound, but
/// one that Silent Mode/a blocking Focus can still suppress the way a
/// true Alarm wouldn't be.
public enum ScheduledNotifications {
    /// Call once at app launch on both platforms. Safe to call every
    /// launch -- UNUserNotificationCenter itself is idempotent about
    /// repeated authorization requests once the user has answered once.
    public static func requestAuthorizationIfNeeded() {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { _, _ in }
    }

    /// `fireAt` is a naive local wall-clock date-time string from the
    /// backend (e.g. "2026-09-19T06:00:00"), computed there against the
    /// real `datetime.now()` in the backend Mac's own local timezone --
    /// which matches Joshua's own local time in the single-user,
    /// single-timezone case this app is built for. Silently does nothing
    /// if the string fails to parse or the time has already passed --
    /// the backend itself already validates this before ever sending the
    /// sentinel, so either case here means the message was corrupted in
    /// transit, not a normal user-facing error to surface.
    public static func schedule(title: String, body: String, fireAt: String) {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss"
        formatter.timeZone = .current
        guard let date = formatter.date(from: fireAt) else { return }
        let interval = date.timeIntervalSinceNow
        guard interval > 0 else { return }

        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: interval, repeats: false)
        let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: trigger)
        UNUserNotificationCenter.current().add(request)
    }
}

File: backend/app/scheduled_notifications.py
"""
Real future-dated device notifications (2026-09-18) -- the closest
working substitute for "set an alarm" or "remind me at X." No third-party
app can create an actual system Alarm on iOS or macOS (a real Apple
platform restriction, not a P Corp scoping choice); this instead tells
whichever client is connected when Frank processes the request to
register a real UNUserNotificationCenter notification with the OS itself
(PCorpKit/ScheduledNotifications.swift), which then fires even if that
app isn't open or connected at the target time.

Same "regular" permission tier as save_memory/set_focus_objective
(SECURITY.md): local-only, reversible (nothing persists server-side --
once the sentinel is sent, the OS on the client device owns it), no
external effect.

Deliberately takes hour/minute/day_offset rather than an absolute ISO
datetime: Claude has no reliable way to know "today's real date" on its
own, so asking it to compute an absolute calendar date itself risks a
hallucinated one. Python computes the real fire_at against its own
datetime.now() instead, using only the small, low-risk pieces (time of
day, day offset) Claude actually needs to interpret from the user's
words.
"""

import json
from datetime import datetime, timedelta

SCHEDULE_NOTIFICATION_TOOL = {
    "name": "schedule_notification",
    "description": (
        "Schedule a real device notification (banner + sound) to fire at a specific future time on whichever "
        "device Joshua is talking to you on right now -- the closest real substitute for 'set an alarm' or "
        "'remind me at X', since no third-party app (including this one) can create an actual system Alarm on "
        "iOS or macOS. Registers directly with that device's own OS notification scheduler, so it fires even if "
        "the app isn't open or connected to you at the target time. Use this when Joshua asks to be woken up, "
        "reminded, or pinged at a specific time -- not for immediate confirmations, which are just a normal "
        "chat reply."
    ),
    "input_schema": {
        "type": "object",
        "properties": {
            "title": {"type": "string", "description": "Notification title, e.g. \"Wake up\" or \"Call reminder\"."},
            "body": {"type": "string", "description": "Notification body text, e.g. \"It's 6:00 AM.\""},
            "hour": {"type": "integer", "description": "Hour of day to fire at, 24-hour format, 0-23 (6 = 6am, 18 = 6pm)."},
            "minute": {"type": "integer", "description": "Minute of the hour, 0-59. Use 0 if Joshua didn't specify one."},
            "day_offset": {
                "type": "integer",
                "description": (
                    "0 = today, 1 = tomorrow, 2 = the day after, etc. If Joshua names a time with no day (e.g. "
                    "\"wake me up at 6am\") and that time has already passed today, use 1, not 0."
                ),
            },
        },
        "required": ["title", "body", "hour", "minute", "day_offset"],
    },
}

SCHEDULED_NOTIFICATION_TOOLS = [SCHEDULE_NOTIFICATION_TOOL]
SCHEDULED_NOTIFICATION_TOOL_NAMES = {tool["name"] for tool in SCHEDULED_NOTIFICATION_TOOLS}


async def execute_scheduled_notification_tool_call(name: str, tool_input: dict, websocket) -> str:
    if name != "schedule_notification":
        return f"Unknown tool: {name}"

    hour = tool_input["hour"]
    minute = tool_input["minute"]
    day_offset = tool_input["day_offset"]

    if not (0 <= hour <= 23) or not (0 <= minute <= 59):
        return f"Invalid time of day: {hour:02d}:{minute:02d} -- hour must be 0-23, minute 0-59."
    if day_offset < 0:
        return "day_offset can't be negative -- can't schedule a notification in the past."

    fire_at = (datetime.now() + timedelta(days=day_offset)).replace(hour=hour, minute=minute, second=0, microsecond=0)
    if fire_at <= datetime.now():
        return (
            f"{fire_at.strftime('%A, %b %d at %I:%M %p')} has already passed -- "
            f"try day_offset={day_offset + 1} if you meant the next day."
        )

    payload = json.dumps({"title": tool_input["title"], "body": tool_input["body"], "fire_at": fire_at.isoformat()})
    await websocket.send_text(f"\n[schedule_notification]{payload}")
    return f"Scheduled {tool_input['title']!r} for {fire_at.strftime('%A, %b %d at %I:%M %p')}."
